src/crypto/crypto_context.cc

```diff
@@ -188,28 +189,66 @@ int SSL_CTX_use_certificate_chain(SSL_CTX* ctx,
                                        issuer);
 }
 
+unsigned long LoadCertsFromFile(  // NOLINT(runtime/int)
+    std::vector<X509*>* certs,
+    const char* file) {
+  ClearErrorOnReturn clear_error_on_return;
+
+  BIOPointer bio(BIO_new_file(file, "r"));
+  if (!bio) return ERR_get_error();
+
+  while (X509* x509 = PEM_read_bio_X509(
+             bio.get(), nullptr, NoPasswordCallback, nullptr)) {
+    certs->push_back(x509);
+  }
+
+  unsigned long err = ERR_peek_error();  // NOLINT(runtime/int)
+  // Ignore error if its EOF/no start line found.
+  if (ERR_GET_LIB(err) == ERR_LIB_PEM &&
+      ERR_GET_REASON(err) == PEM_R_NO_START_LINE) {
+    return 0;
+  }
+
+  return err;
+}
+
 }  // namespace
 
-X509_STORE* NewRootCertStore() {
+X509_STORE* NewRootCertStore(Environment* env) {
   static std::vector<X509*> root_certs_vector;
+  static bool root_certs_vector_loaded = false;
   static Mutex root_certs_vector_mutex;
   Mutex::ScopedLock lock(root_certs_vector_mutex);
 
-  if (root_certs_vector.empty() &&
-      per_process::cli_options->ssl_openssl_cert_store == false) {
-    for (size_t i = 0; i < arraysize(root_certs); i++) {
-      X509* x509 =
-          PEM_read_bio_X509(NodeBIO::NewFixed(root_certs[i],
-                                              strlen(root_certs[i])).get(),
-                            nullptr,   // no re-use of X509 structure
-                            NoPasswordCallback,
-                            nullptr);  // no callback data
+  if (!root_certs_vector_loaded) {
+    if (per_process::cli_options->ssl_openssl_cert_store == false) {
+      for (size_t i = 0; i < arraysize(root_certs); i++) {
+        X509* x509 = PEM_read_bio_X509(
+            NodeBIO::NewFixed(root_certs[i], strlen(root_certs[i])).get(),
+            nullptr,  // no re-use of X509 structure
+            NoPasswordCallback,
+            nullptr);  // no callback data
+
+        // Parse errors from the built-in roots are fatal.
+        CHECK_NOT_NULL(x509);
 
-      // Parse errors from the built-in roots are fatal.
-      CHECK_NOT_NULL(x509);
+        root_certs_vector.push_back(x509);
+      }
+    }
 
-      root_certs_vector.push_back(x509);
+    if (!extra_root_certs_file.empty()) {
+      unsigned long err = LoadCertsFromFile(  // NOLINT(runtime/int)
+          &root_certs_vector,
+          extra_root_certs_file.c_str());
+      if (err) {
+        ProcessEmitWarning(env,
+                           "Ignoring extra certs from `%s`, load failed: %s\n",
+                           extra_root_certs_file.c_str(),
+                           ERR_error_string(err, nullptr));
+      }
     }
+
+    root_certs_vector_loaded = true;
```

Comment:
I'm not sure why this boolean is needed. Is it to allow an empty set of root certificates? Why?

Reply:
If `per_process::cli_options->ssl_openssl_cert_store == true` and `NODE_EXTRA_CA_CERTS` isn't specified, `root_certs_vector` will remain empty. 

The inner if statements within the block will still properly gate everything and avoid any issues, but I opted for the boolean because it's a bit more robust - I wanted to make it clear to anyone modifying that code that having `root_certs_vector` empty but "initialized" is a valid scenario.

Reply:
When `NODE_EXTRA_CA_CERTS` specified but pointed at a missing or invalid file, the message `Warning: Ignoring extra certs from '%s', load failed: %s` will be printed to `stderr`.

If this happens and `per_process::cli_options->ssl_openssl_cert_store` is true, the `root_certs_vector` will be empty, causing the the warning message will be printed on every call to `NewRootCertStore`. The boolean check resolves this edge case.